omod/src/main/java/org/openmrs/module/referencedemodata/web/controller/DemoDataGenerationController.java

```diff
@@ -0,0 +1,96 @@
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.v1_0.controller.BaseRestController;
+import org.openmrs.scheduler.TaskDefinition;
+import org.openmrs.scheduler.tasks.AbstractTask;
+import org.openmrs.util.PrivilegeConstants;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/" + ReferenceDemoDataConstants.REFERENCE_DEMO_DATA_URI)
+public class DemoDataGenerationController extends BaseRestController {
+	
+	private static ReferenceDemoDataActivator referenceDemoDataActivator;
+	
+	private static final String REFERENCE_DEMO_DATA_TASK_NAME = "Reference demo data generation task";
+	
+	public static final String NUMBER_OF_DEMO_PATIENTS_PARAMETER = "numberOfDemoPatients";
+
+	@RequestMapping(value = ReferenceDemoDataConstants.GENERATE_DEMO_DATA_URI, method = RequestMethod.GET, produces = "application/json")
+	@ResponseBody()
+	public String generateDemoData(HttpServletRequest request) throws Exception {
+		
+		String numberOfDemoPatients = request.getParameter(NUMBER_OF_DEMO_PATIENTS_PARAMETER);
+		
+		try {
+			Integer.parseInt(numberOfDemoPatients);
+		}
+		catch (NumberFormatException e) {
+			String errorMsg = String.format("Could not parse '%s' as an integer", numberOfDemoPatients);
+			return String.format("{\"error\": \"%s\"}", errorMsg);
+		}
+		
+		Context.addProxyPrivilege(PrivilegeConstants.MANAGE_SCHEDULER);
+		TaskDefinition taskDef = Context.getSchedulerService().getTaskByName(REFERENCE_DEMO_DATA_TASK_NAME);
+		if (taskDef == null) {
+			taskDef = new TaskDefinition();
+			taskDef.setTaskClass(GenerateDemoDataTask.class.getName());
+			taskDef.setStartOnStartup(false);
+			taskDef.setStartTime(null); // Executes immediately
+			taskDef.setName(REFERENCE_DEMO_DATA_TASK_NAME);
+			taskDef.setDescription(REFERENCE_DEMO_DATA_TASK_NAME);
+			taskDef.setRepeatInterval(0L); // Executes once
+			taskDef.setProperty("numberOfDemoPatients", numberOfDemoPatients);
+			Context.getSchedulerService().saveTaskDefinition(taskDef);
+		}
+		if (Context.getPatientService().getAllPatients().size() < Integer.parseInt(numberOfDemoPatients)) {
+			int remainingNumberOfPatientsToGenerate =  Integer.parseInt(numberOfDemoPatients) - Context.getPatientService().getAllPatients().size();
+			Context.addProxyPrivilege(PrivilegeConstants.MANAGE_GLOBAL_PROPERTIES);
+			GlobalProperty createDemoPatients = new GlobalProperty(
+					ReferenceDemoDataConstants.CREATE_DEMO_PATIENTS_ON_NEXT_STARTUP, "" + Integer.toString(remainingNumberOfPatientsToGenerate));
+			Context.getAdministrationService().saveGlobalProperty(createDemoPatients);
+			Context.removeProxyPrivilege(PrivilegeConstants.MANAGE_GLOBAL_PROPERTIES);
+			synchronized (new Integer(1)) {
```

Comment:
`synchronized (new Integer(1))` doesn't actually work; the `Integer()` constructor always returns a new instance, so you're effectively synchronizing on a local variable; since no other thread can see this local variable, this synchronization block doesn't actually synchronize anything. (`Integer.valueOf(1)` would work, but that's because the JVM pre-caches instances for a few values).

I'm not quite certain why this is here at all. If access to the `scheduleTask()` function needs to be synchronized, that's the responsibility of that function.

Reply:
At the moment `scheduleTask()` does not handle synchronisation. I guess I can handle this here along the lines,

```
synchronized (DemoDataGenerationController.class) {
	Context.getSchedulerService().scheduleTask(taskDef);
}
```

Reply:
Fair enough... My point was was just that what's here is equivalent to no synchronization but with extra steps. Synchronizing on the controller class makes sense.